cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/DefaultOptimizationOptionsGenerator.java

```diff
@@ -17,22 +17,30 @@
   public OptimizationOptions optimizationOptionsForGoalViolationDetection(ClusterModel clusterModel,
                                                                           Set<String> excludedTopics,
                                                                           Set<Integer> excludedBrokersForLeadership,
-                                                                          Set<Integer> excludedBrokersForReplicaMove) {
+                                                                          Set<Integer> excludedBrokersForReplicaMove,
+                                                                          Set<String> mustHaveLeaderReplicaPerBrokerTopics,
+                                                                          int minLeaderReplicaCountForTopicsMustHaveLeaderReplicaPerBroker) {
     return new OptimizationOptions(excludedTopics,
                                    excludedBrokersForLeadership,
                                    excludedBrokersForReplicaMove,
-                                   true);
+                                   true,
+                                   mustHaveLeaderReplicaPerBrokerTopics,
+                                   minLeaderReplicaCountForTopicsMustHaveLeaderReplicaPerBroker);
   }
 
   @Override
   public OptimizationOptions optimizationOptionsForCachedProposalCalculation(ClusterModel clusterModel,
-                                                                             Set<String> excludedTopics) {
+                                                                             Set<String> excludedTopics,
+                                                                             Set<String> mustHaveLeaderReplicaPerBrokerTopics,
+                                                                             int minLeaderReplicaCountForTopicsMustHaveLeaderReplicaPerBroker) {
     return new OptimizationOptions(excludedTopics,
                                    Collections.emptySet(),
                                    Collections.emptySet(),
                                    false,
                                    Collections.emptySet(),
-                                   false);
+                                   false,
+                                   mustHaveLeaderReplicaPerBrokerTopics,
+                                   minLeaderReplicaCountForTopicsMustHaveLeaderReplicaPerBroker);
```

Comment:
Created an [issue](https://github.com/linkedin/cruise-control/issues/1432) to keep track of your second point.

Reply:
Moved these two variables to `BalancingConstraint`. Btw, initially, I put them in `BalancingConstraint` and you told me to put it somewhere else during an offline discussion. So, I decided to put them in `OptimizationOptions` since there is no other suitable place. But anyway, I agree that putting them in `BalancingConstraint` is better.